replication-api-impl/src/main/java/com/connexta/replication/api/impl/data/AbstractPersistable.java

```diff
@@ -0,0 +1,178 @@
+/**
+ * Copyright (c) Connexta
+ *
+ * <p>This is free software: you can redistribute it and/or modify it under the terms of the GNU
+ * Lesser General Public License as published by the Free Software Foundation, either version 3 of
+ * the License, or any later version.
+ *
+ * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details. A copy of the GNU Lesser General Public
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package com.connexta.replication.api.impl.data;
+
+import com.connexta.ion.replication.api.ReplicationPersistenceException;
+import com.connexta.replication.api.data.Persistable;
+import com.connexta.replication.api.impl.persistence.pojo.Pojo;
+import com.google.common.annotations.VisibleForTesting;
+import java.util.UUID;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+import javax.annotation.Nullable;
+
+/**
+ * This class defines some data that any object to be saved in persistence should have and methods
+ * to manipulate that data.
+ *
+ * @param <P> the type of pojo associated with this persistable
+ */
+public abstract class AbstractPersistable<P extends Pojo> implements Persistable {
+  /** A string representing the type of this object. Used when generating exception or logs. */
+  protected final String type;
+
+  private String id;
+
+  protected AbstractPersistable(String type) {
+    this.type = type;
+    this.id = UUID.randomUUID().toString();
+  }
+
+  protected AbstractPersistable(String type, @Nullable String id) {
+    this.type = type;
+    this.id = id;
+  }
+
+  @Override
+  public String getId() {
+    return id;
+  }
+
+  /**
+   * Serializes this object to the specified pojo that can be serialized to a spring-data supported
+   * data store. Any implementation of this method in a subclass should first make a call to the
+   * super version of this method before serializing its own properties.
+   *
+   * @param pojo the pojo to serialize this object to
+   * @return <code>pojo</code> for chaining
+   * @throws ReplicationPersistenceException if an error occurs while trying to serialize this
+   *     object
+   */
+  protected P writeTo(P pojo) {
```

Comment:
❓ Do we need to set the version here?

Reply:
I am on the fence. I am assuming that we are always serializing to the latest version which is expected to be set by the constructor of the pojo. On one side, I am not a fan on relying on something else to do it. On the other side, I didn't want the persistable to be aware of the version. But since they are aware since they are responsible for converting, I will introduce it in each specific persistable `writeToPojo)_` classes.